src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2221,27 +2221,28 @@ public override BoundNode VisitConditionalAccess(BoundConditionalAccess node)
                 LearnFromNonNullTest(receiver, ref this.State);
             }
 
-            var accessExpressionType = VisitRvalueWithState(node.AccessExpression);
+            var accessTypeWithAnnotations = VisitLvalueWithAnnotations(node.AccessExpression);
+            TypeSymbol accessType = accessTypeWithAnnotations.Type;
             Join(ref this.State, ref receiverState);
 
-            // Per LDM 2019-02-13 decision, the result of a conditional access might be null even if
-            // both the receiver and right-hand-side are believed not to be null.
-            NullableFlowState resultState = NullableFlowState.MaybeNull;
-
-            // https://github.com/dotnet/roslyn/issues/29956 Use flow analysis type rather than node.Type
-            // so that nested nullability is inferred from flow analysis. See VisitConditionalOperator.
-            TypeSymbol type = node.Type;
-
             // If the result type does not allow annotations, then we produce a warning because
             // the result may be null.
-            if (RequiresSafetyWarningWhenNullIntroduced(type))
+            if (RequiresSafetyWarningWhenNullIntroduced(accessType))
             {
-                ReportSafetyDiagnostic(ErrorCode.WRN_ConditionalAccessMayReturnNull, node.Syntax, node.Type);
+                ReportSafetyDiagnostic(ErrorCode.WRN_ConditionalAccessMayReturnNull, node.Syntax, accessType);
             }
 
+            var resultType =
+                accessType.SpecialType == SpecialType.System_Void || accessType.IsPointerType() || accessType.IsRestrictedType() ? compilation.GetSpecialType(SpecialType.System_Void) :
+                accessType.CanBeAssignedNull() ? accessType :
+                accessType.IsNonNullableValueType() ? compilation.GetSpecialType(SpecialType.System_Nullable_T).Construct(ImmutableArray.Create(accessTypeWithAnnotations)) :
+                compilation.GetSpecialType(SpecialType.System_Void);
```

Comment:
>compilation.GetSpecialType(SpecialType.System_Void); [](start = 16, length = 52)

This logic is way too complicated and it looks like it is trying to duplicate similar logic from the binding layer. Either we should extract the logic and share it between the two layers, or use much simple logic here that is simply based off of the type that the original node has:
- If original type is void -> void
- if original type is ```Nullable<T>`` and ```accessType``` is not -> ```Nullable<accessType>```
- Otherwise accessType

We should also add an assert that the type we get at the end is equal to the original type, ignoring all possible aspects.

Reply:
Unfortunately, the code in the binder is intertwined with error reporting and other error recovery in a way that makes sharing it nontrivial.  Like most of the nullable walker, I believe this code simply has to reproduce the correct language semantics.  I did try a simpler logic, but it is wrong.

---
In reply to: [276447420](https://github.com/dotnet/roslyn/pull/34973#discussion_r276447420) [](ancestors = 276447420)

Reply:
I misread your suggestion.  I'll try that.

---
In reply to: [278341199](https://github.com/dotnet/roslyn/pull/34973#discussion_r278341199) [](ancestors = 278341199,276447420)

Reply:
The suggested assertion appears in `VisitExpressionWithoutStackGuard`, which is on the path from the caller.

---
In reply to: [278345603](https://github.com/dotnet/roslyn/pull/34973#discussion_r278345603) [](ancestors = 278345603,278341199,276447420)